predicators/pybullet_helpers/ikfast/utils.py

```diff
@@ -0,0 +1,325 @@
+    ik_joints = [joint_info.jointIndex for joint_info in ik_joint_infos]
+
+    tool_link = robot.tool_link_id
+
+    # Get the desired pose of the end-effector in the base frame
+    base_from_ee = get_base_from_ee(robot, tool_link, world_from_target)
+    position = list(base_from_ee.position)
+    rot_matrix = matrix_from_quat(base_from_ee.orientation).tolist()
+
+    # Sampler for free joints
+    generator = free_joints_generator(robot, free_joint_infos, max_distance)
+    if max_attempts < np.inf:  # pragma: no cover
+        generator = islice(generator, max_attempts)
+
+    difference_fn = get_difference_fn(ik_joint_infos)
+    current_conf = get_joint_positions(robot.robot_id, ik_joints,
+                                       robot.physics_client_id)
+
+    start_time = time.perf_counter()
+    for free_positions in generator:
+        elapsed_time = time.perf_counter() - start_time
+        if elapsed_time >= max_time:
+            logging.warning("Max time reached. No IKFast solution found.")
+            break
+
+        # Call IKFast to compute candidates for sampled free joint positions
+        ik_candidates: Optional[List[
+            List[float]]] = ikfast.get_ik(  # type: ignore
+                rot_matrix, position, list(free_positions))
+        if ik_candidates is None:
+            continue
+
+        # Shuffle the candidates to avoid any biases
+        random.shuffle(ik_candidates)
+
+        # Check candidates are valid
+        for conf in ik_candidates:
+            difference = difference_fn(current_conf, conf)
+            if not violates_joint_limits(ik_joint_infos, conf) and (
+                    np.linalg.norm(difference, ord=norm) <= max_distance):
+                yield conf
+
+
+def ikfast_closest_inverse_kinematics(
+        robot: SingleArmPyBulletRobot,
+        world_from_target: Pose) -> List[JointPositions]:
+    """Runs IKFast and returns the solutions sorted in order of closets
+    distance to the robot's current joint positions.
+
+    Parameters
+    ----------
+    robot: SingleArmPyBulletRobot
+    world_from_target: target pose of the tool link in the world frame
+
+    Returns
+    -------
+    A list of joint states that satisfy the given arguments.
+    If no solutions are found, an empty list is returned.
+    """
+    start_time = time.perf_counter()
+    ik_joint_infos, _ = get_ikfast_joints(robot)
+    ik_joints = [joint_info.jointIndex for joint_info in ik_joint_infos]
+    current_conf = get_joint_positions(robot.robot_id, ik_joints,
+                                       robot.physics_client_id)
+
+    generator = ikfast_inverse_kinematics(
+        robot,
+        world_from_target,
+        max_time=CFG.ikfast_max_time,
+        max_distance=CFG.ikfast_max_distance,
+        max_attempts=CFG.ikfast_max_attempts,
+        norm=CFG.ikfast_norm,
+    )
+
+    # Only use up to the max candidates specified
+    if CFG.ikfast_max_candidates < np.inf:  # pragma: no cover
```

Comment:
Good point - right now the generator will stop generating once it reaches max time, max attempts (and maybe some other ones I can't remember). 

The max time is set to like 0.05 seconds, but if that is infinity as well then we get an infinite loop. Would be worth validating the settings so make sure this would terminate at some point.

Reply:
Added check that one of max time, max attempts or max candidates is < np.inf